packages/docprovider/src/yprovider.ts

```diff
@@ -41,17 +44,28 @@ export class WebSocketProviderWithLocks
     this._path = options.path;
     this._contentType = options.contentType;
     this._serverUrl = options.url;
-    const color = '#' + env.getParam('--usercolor', getRandomColor().slice(1));
-    const name = env.getParam('--username', getAnonymousUserName());
+
     const awareness = options.ymodel.awareness;
     const currState = awareness.getLocalState();
-    // only set if this was not already set by another plugin
-    if (currState && currState.name == null) {
-      options.ymodel.awareness.setLocalStateField('user', {
-        name,
-        color
-      });
-    }
+    let color = '#' + env.getParam('--usercolor', getRandomColor().slice(1));
+    let name = env.getParam('--username', getAnonymousUserName());
+    const state = options.state;
+    const user = state.fetch(USER);
+    user.then(param => {
+      if (param === undefined) {
+        state.save(USER, `${name},${color}`);
```

Comment:
Maybe store it as an object, e.g. `{ name: name, color: color }`? I think this is supported by StateDB as in tests:

https://github.com/jupyterlab/jupyterlab/blob/dbdefed9db9332381fe4104bdf53ec314451951f/packages/statedb/test/statedb.spec.ts#L162-L171

Reply:
Thanks for the suggestion @krassowski, I tried storing as an object but did not find an easy way to deconstruct. Any pointer to an example how ReadonlyJSONValue should be deconstructed would be much appreciated.

Reply:
You would first define an interface for the stored values extending e.g. `PartialJSONObject` and ten coerce the type on read using `as`. For an example interface see [`Private.ILayout`](https://github.com/jupyterlab/jupyterlab/blob/e9d8490d154d801d86722bb7111fa5b2eb5ab2d0/packages/application/src/layoutrestorer.ts#L506) of layout restored, and for destructing see its [`fetch`](https://github.com/jupyterlab/jupyterlab/blob/e9d8490d154d801d86722bb7111fa5b2eb5ab2d0/packages/application/src/layoutrestorer.ts#L176-L182).

Reply:
Thank you:)  looks like [this](https://github.com/jupyterlab/jupyterlab/pull/10490/commits/a7b859ee1514012c36ce59b9cad37ce8b4f07fe3) worked. It turns out the variable names, when deconstructing, need to match the property names exactly ...